### PR TITLE
UUID callback functionality

### PR DESCRIPTION
**Pull request by scolear**

### CODE REVIEW

**Comment by scolear** [comment]
Shouldn't forget to set the mocknet-specific things back to production-state before merging.

**Comment by scolear** [comment]
https://github.com/DLC-link/oracle-interface/pull/10

**Comment by scolear** [comment]
To do: 
- [ ] Update readme with new functionality
- [ ] Update & add new tests

File: contracts/dlc-create-callback-trait.clar [hunk 1]
```diff
@@ -0,0 +1,5 @@
+(define-trait dlc-create-callback-trait
+    (
+        (post-create-handler (uint (buff 8)) (response bool uint))
```
Comment (scolear) [inline contracts/dlc-create-callback-trait.clar:3]: post-dlc-create-handler

File: contracts/dlc-manager-pricefeed.clar [hunk 2]
```diff
@@ -144,13 +158,37 @@
...
+  )
+)
+
+(define-public (register-contract (contract-address <cb-trait>))
+  (begin
+    (asserts! (is-eq contract-owner tx-sender) err-not-contract-owner)
+    (print { 
+      contract-address: contract-address,
+      event-source: "dlclink:register-contract:v1" })
+    (nft-mint? registered-contract (contract-of contract-address) tx-sender)
```
Comment (scolear) [inline contracts/dlc-manager-pricefeed.clar:178]: put dlc-manager principal as target

File: contracts/dlc-manager-pricefeed.clar [hunk 3]
```diff
@@ -144,13 +158,37 @@
...
+    (print { 
+      contract-address: contract-address,
+      event-source: "dlclink:unregister-contract:v1" })
+    (nft-burn? registered-contract (contract-of contract-address) tx-sender)
+  )
+)
+
+(define-read-only (is-contract-registered (contract-address principal))
+  (is-some (nft-get-owner? registered-contract contract-address))
 )
```
Comment (scolear) [inline contracts/dlc-manager-pricefeed.clar:194]: New function: is-dlc-owner to check

File: contracts/dlc-manager-pricefeed.clar [hunk 4]
```diff
@@ -88,7 +99,8 @@
     (print {
```
Comment (scolear) [inline contracts/dlc-manager-pricefeed.clar:99]: swap lines 96-97

Comment (scolear) [reply contracts/dlc-manager-pricefeed.clar:99]: Auth checking should be on top

File: contracts/dlc-manager-pricefeed.clar [hunk 5]
```diff
@@ -14,32 +14,58 @@
...
-;;opens a new dlc
-(define-public (create-dlc-internal (uuid (buff 8)) (asset (buff 32)) (closing-time uint) (emergency-refund-time uint) (creator principal))
+;;emits an event - see README for more details
+;;asset: the ticker symbol for the asset that will be used to closing price data, e.g. 'btc' (https://github.com/redstone-finance/redstone-api/blob/main/docs/ALL_SUPPORTED_TOKENS.md)
+;;strike-price: the over/under price for the DLC bet
+;;closing-time: the UTC time in seconds after which the contract can be closed, and will fetch the closing price
+;;emergency-refund-time: the time at which the DLC will be available for refund
+;;the calling contract's name
+;;nonce provided for the dlc by the protocol-contract
+(define-public (create-dlc (asset (buff 32)) (strike-price uint) (closing-time uint) (emergency-refund-time uint) (callback-contract principal) (nonce uint))
```
Comment (scolear) [inline contracts/dlc-manager-pricefeed.clar:53]: callback-contract is cb-trait?

File: contracts/external/protocol-contract.clar [hunk 6]
```diff
@@ -0,0 +1,90 @@
+
+;; protocol-contract
```
Comment (scolear) [inline contracts/external/protocol-contract.clar:2]: Let's call it sample-protocol-contract

File: contracts/external/protocol-contract.clar [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            nonce: nonce,
+            active: true
+        }))
+        (ok true)
+    )
+)
+
+;; public functions
+;;
+(define-public (setup-user-contract (asset (buff 32)) (strike-price uint) (closing-time uint) (emergency-refund-time uint))
```
Comment (scolear) [inline contracts/external/protocol-contract.clar:60]: Add comments

File: scripts/redstone-tests.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  }
+}
+
+(async () => {
+
+  interface PriceDataWithSig extends PriceData {
+    liteEvmSignature: string;
+  }
+
+  // const dataPackageNow = await redstoneExtended.oracle.getFromDataFeed("redstone", assetName);
```
Comment (scolear) [inline scripts/redstone-tests.ts:60]: Comment, cleanup, or remove

File: tests/discreet-log-storage_test.ts [hunk 9]
```diff
@@ -11,7 +11,7 @@ import type { PricePackage, Block } from "./deps.ts";
 const BTChex = "BTC";
```
Comment (scolear) [inline tests/discreet-log-storage_test.ts:11]: Write new tests and update

File: contracts/dlc-manager-pricefeed.clar (hunk 5 above)
Comment (scolear) [reply contracts/dlc-manager-pricefeed.clar:53]: Unfortunately setting it to <cb-trait> makes the sample-protocol-contract err out on calling this function.

File: tests/discreet-log-storage_test.ts [hunk 10]
```diff
@@ -151,47 +154,45 @@ Clarinet.test({
...
+      block.receipts[0].result.expectOk().expectBool(true);
+      const event = block.receipts[0].events[0];
 
       assertEquals(typeof event, 'object');
       assertEquals(event.type, 'contract_event');
       assertEquals(event.contract_event.topic, "print");
-      assertStringIncludes(event.contract_event.value, "uuid: 0x66616b6575756964")
-      assertStringIncludes(event.contract_event.value, "asset: 0x425443")
+      assertStringIncludes(event.contract_event.value, "uuid: 0x66616b6575756964");
+      assertStringIncludes(event.contract_event.value, "asset: 0x425443");assertStringIncludes(event.contract_event.value, 'event-source: "dlclink:close-dlc:v1"');
```
Comment (scolear) [inline tests/discreet-log-storage_test.ts:172]: Missing newline

File: tests/discreet-log-storage_test.ts [hunk 11]
```diff
@@ -75,71 +91,58 @@ Clarinet.test({
...
+        const dlc: any = block.receipts[0].result.expectSome().expectTuple();
 
         assertEquals(hex2ascii(dlc.asset), "BTC");
         assertEquals(hex2ascii(dlc.uuid), "fakeuuid");
         assertEquals(dlc["closing-price"], "none");
-        assertEquals(dlc.creator, wallet_1.address);
-    },
-});
-
-Clarinet.test({
```
Comment (scolear) [inline tests/discreet-log-storage_test.ts:120]: We probably need this test back.

**Review by sosaucily — COMMENTED** [review]
I see there is a file called deps.ts under the tests directory. Is this only for tests, or are those functions used in the prod code as well? Better for anything under the tests dir to only be used for testing.

**Review by sosaucily — COMMENTED** [review]
Also i see a duplicate of the hex2ascii function in both the test file and prod. Better to import a helpers file, and just the prod one, not duplicate.

For functions that need to be used for tests, can import from a test-helpers.ts file. But only if those functions are test only.

I think this is the best way to go, but let me know if you think differently.

**Comment by sosaucily** [comment]
Great! On the testing side, i know it was a pretty big job to get them all passing. Big structural changes, which are looking good. Good job.

In the future, when we're adding the enhanced closing capabilities as part of the next story, I think we should
1. Clean up these tests, so they're in a more sensible order, and in context blocks for logical organization.
2. Add more error cases
3. Add more around closing 

But this is definitely good enough for this PR. Great job dude. Just a few comments above about file structure and G2G

**Comment by scolear** [comment]
> Also i see a duplicate of the hex2ascii function in both the test file and prod. Better to import a helpers file, and just the prod one, not duplicate.
> 
> For functions that need to be used for tests, can import from a test-helpers.ts file. But only if those functions are test only.
> 
> I think this is the best way to go, but let me know if you think differently.

I think it's nice that the tests folder has only and everything it needs to run the tests, like deps.ts. I guess we can move the hex2ascii function in there too. It only appears once more, in a very specific script (get-all-open-dlcs.ts), which I think don't warrant creating a whole new helper-functions file just for that one function. LMK what you think

**Comment by sosaucily** [comment]
good 'nuff! nice job, lgtm

**Review by sosaucily — APPROVED** [review]

**Pull request merged**